File: python/today3/spiral_matrix.py
# https://leetcode.com/problems/spiral-matrix/description/
from typing import List

class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        left = 0
        right = len(matrix[0]) - 1
        top = 0
        bottom = len(matrix) - 1
        
        size = len(matrix) * len(matrix[0])
        result = []

        while len(result) < size:
            for col in range(left, right+1):
                result.append(matrix[top][col])
                if len(result) == size:
                    return result
            top += 1
            for row in range(top, bottom+1):
                result.append(matrix[row][right])
                if len(result) == size:
                    return result
            right -= 1
            for col in range(right, left-1, -1):
                result.append(matrix[bottom][col])
                if len(result) == size:
                    return result
            bottom -= 1
            for row in range(bottom, top-1, -1):
                result.append(matrix[row][left])
                if len(result) == size:
                    return result
            left += 1
        return result

File: typescript/solved/spiral-matrix.ts
/**
 * https://leetcode.com/problems/spiral-matrix/
 */
export function spiralOrder(matrix: number[][]): number[] {
  let topBound = 0;
  let bottomBound = matrix.length - 1;
  let leftBound = 0;
  let rightBound = matrix[0].length - 1;

  const cellCount = matrix.length * matrix[0].length;
  const result: number[] = [];

  while (result.length < cellCount) {
    // Traverse right
    for (let i = leftBound; i <= rightBound; i++) {
      result.push(matrix[topBound][i]);
    }
    if (result.length >= cellCount) return result;
    topBound++;

    // Traverse down
    for (let i = topBound; i <= bottomBound; i++) {
      result.push(matrix[i][rightBound]);
    }
    if (result.length >= cellCount) return result;
    rightBound--;

    // Traverse left
    for (let i = rightBound; i >= leftBound; i--) {
      result.push(matrix[bottomBound][i]);
    }
    if (result.length >= cellCount) return result;
    bottomBound--;

    // Traverse up
    for (let i = bottomBound; i >= topBound; i--) {
      result.push(matrix[i][leftBound]);
    }
    leftBound++;
  }

  return result;
}
